doc/specs/#642 - Buffer Exporting and Logging/#642 - Buffer Exporting and Logging.md

```diff
@@ -0,0 +1,380 @@
+* **Story F**: The profile "bar" is configured to automatically log when opened.
+* **Story G**: This is the action bound to <kbd>f4</kbd>.
+
+In addition,
+* When opening the profile "foo", it will not automatically log to a file.
+  - Pressing <kbd>f3</kbd> will begin logging to `c:\foo.txt`
+  - Pressing <kbd>f4</kbd> will begin logging to `c:\logs\foo.log`
+
+### Path formatting
+
+[TODO!]: # TODO!
+
+For discussion: What syntax do we want?
+* PuTTY uses `&Y`, `&M`, `&D`, `&T`, `&H`, `&P` for year, month, day, time, host
+  and port respectively.
+* SecureCRT uses:
+  - `%H` – hostname
+  - `%S` – session name
+  - `%Y` – four-digit year
+  - `%M` – two-digit month
+  - `%D` – two-digit day of the month
+  - `%h` – two-digit hour
+  - `%m` – two-digit minute
+  - `%s` – two-digit seconds
+  - `%t` – three-digit milliseconds
+  - `%%` – percent (%)
+  - `%envvar%` – environment variable (for instance `%USERNAME%`)
+
+We have some precedent for formatting with `${braces}`, a la the iterable
+command in the Command Palette (e.g `${profile.name}`). Additionally, [#9287]
+implements support for environment variables in the Terminal with the
+`${env:VARIABLE}` syntax.
+
+What variables do we want exposed, and how do we want users to be able to format
+them?
+
+This doc was initially authored assuming we'd go with a `${braces}` syntax, like:
+  - `${profile}` – profile name
+  - `${year}` – four-digit year
+  - `${month}` – two-digit month
+  - `${day}` – two-digit day of the month
+  - `${hour}` – two-digit hour
+  - `${minute}` – two-digit minute
+  - `${second}` – two-digit second
+  - `${ms}` – three-digit milliseconds
+  - `${env:variable}` – environment variable (for instance `${env:USERPROFILE}`)
+    (inspired by [#9287])
+
+### Exporting vs Logging
+As far as specific implementation details goes, exporting is the easier work to
+do. [#11062] already wires up the `TerminalApp` to retrieve the buffer contents
+from the `TermControl`, so writing them at request is easy.
+
+Logging is harder. We don't want the `TermControl` telling the `TerminalApp`
+layer about every piece of output logged. Especially in the post-[#5000] world
+where that's a cross-process hop. Instead, we'll want the `ControlCore` /
+`ControlInteractivity` to do _logging_ themselves.
+
+### Logging Mechanics
+
+#### When do we log?
+
+[TODO!]: # TODO!
+
+When do we decide to actually log? Take for example typing in a `pwsh` or
+`bash` prompt. Imagine the user types
+<kbd>a</kbd><kbd>s</kbd><kbd>d</kbd><kbd>f</kbd>, then hits
+<kbd>Bksp</kbd><kbd>Bksp</kbd>, such that the prompt is just `as`. What should
+the log contain? `asdf^h ^h^h ^h`<sup>[[1]](#footnote-1)</sup>? `as`?
+
+My worry with logging the backspaces is that conpty is sometimes a bit noisier
+than it needs to be with using `^H` as a cursor positioning sequence. Should we
+only log lines when the cursor newlines or otherwise moves from the line it is
+currently on?

```

Comment:
What happens in `vim` then? Or when performing a `cls`? I guess the mechanics of this we'll have to compare with PuTTY a bit closer.

Reply:
Not sure if I understood the discussion correctly, but we must be able to log everything that happened. Even within the same line, and even if it is noisy. These logs are sometimes required for investigating/debugguing. 

Since there might be too much updates we will need to use some async writes with buffering (and ensure everything is flushed when closing).

Reply:
I guess my real concern is "does backspace count as a printable character". Yes-ish? By itself it's a non-destructive character. But if we don't treat it as printable, then <kbd>a</kbd><kbd>s</kbd><kbd>d</kbd><kbd>f</kbd><kbd>Bksp</kbd><kbd>Bksp</kbd> gets emitted as `asdf^h ^h^h ^h`, but then would get logged as `asdf  ` (with two spaces at the end there). So I guess, backspace should count as printable. 

The "log _everything_" option is frankly just way easier to implement than the "printable characters only" one. Maybe we start with that one 😛